File: packages/go/transport.go
// ---------------------------------------------------------------------------
// uncaught-go — local file transport
// ---------------------------------------------------------------------------
//
// Writes events to the .uncaught/ directory in the same format as the
// TypeScript SDK, enabling cross-language compatibility.
//
// Directory structure:
//
//	.uncaught/
//	  events/
//	    <fingerprint>/
//	      event-<timestamp>.json
//	      latest.json
//	  fix-prompts/
//	    <fingerprint>.md
//	  issues.json
// ---------------------------------------------------------------------------

package uncaught

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
)

// LocalFileTransport writes events to the local .uncaught/ directory.
type LocalFileTransport struct {
	mu      sync.Mutex
	baseDir string
	inited  bool
}

// NewLocalFileTransport creates a new local file transport.
// If baseDir is empty, it defaults to "./.uncaught" relative to the working directory.
func NewLocalFileTransport(baseDir string) *LocalFileTransport {
	if baseDir == "" {
		cwd, err := os.Getwd()
		if err != nil {
			cwd = "."
		}
		baseDir = filepath.Join(cwd, ".uncaught")
	}
	return &LocalFileTransport{
		baseDir: baseDir,
	}
}

// Send writes an event to the .uncaught/ directory.
func (t *LocalFileTransport) Send(event *UncaughtEvent) error {
	t.mu.Lock()
	defer t.mu.Unlock()

	if err := t.init(); err != nil {
		return fmt.Errorf("transport init failed: %w", err)
	}

	fp := event.Fingerprint
	eventDir := filepath.Join(t.baseDir, "events", fp)

	// Ensure event directory exists
	if err := os.MkdirAll(eventDir, 0755); err != nil {
		return fmt.Errorf("mkdir events/%s: %w", fp, err)
	}

	// Marshal event to JSON
	eventJSON, err := json.MarshalIndent(event, "", "  ")
	if err != nil {
		return fmt.Errorf("marshal event: %w", err)
	}

	// Write timestamped event file (atomic: .tmp -> rename)
	ts := strings.NewReplacer(":", "-", ".", "-").Replace(event.Timestamp)
	eventFile := fmt.Sprintf("event-%s.json", ts)
	eventPath := filepath.Join(eventDir, eventFile)
	if err := atomicWrite(eventPath, eventJSON); err != nil {
		return fmt.Errorf("write event file: %w", err)
	}

	// Write / overwrite latest.json
	latestPath := filepath.Join(eventDir, "latest.json")
	if err := atomicWrite(latestPath, eventJSON); err != nil {
		return fmt.Errorf("write latest.json: %w", err)
	}

	// Write fix-prompt Markdown file
	promptFile := fmt.Sprintf("%s.md", fp)
	promptPath := filepath.Join(t.baseDir, "fix-prompts", promptFile)
	if err := atomicWrite(promptPath, []byte(event.FixPrompt)); err != nil {
		return fmt.Errorf("write fix-prompt: %w", err)
	}

	// Update issues.json index
	if err := t.updateIssuesIndex(event, eventFile, promptFile); err != nil {
		return fmt.Errorf("update issues.json: %w", err)
	}

	return nil
}

// Flush is a no-op for the local file transport since writes are synchronous.
func (t *LocalFileTransport) Flush() error {
	return nil
}

// init ensures the directory structure exists and .gitignore is updated.
func (t *LocalFileTransport) init() error {
	if t.inited {
		return nil
	}

	// Create directory structure
	if err := os.MkdirAll(filepath.Join(t.baseDir, "events"), 0755); err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Join(t.baseDir, "fix-prompts"), 0755); err != nil {
		return err
	}

	// Auto-add .uncaught/ to .gitignore
	t.ensureGitignore()

	t.inited = true
	return nil
}

// ensureGitignore adds .uncaught/ to .gitignore if not already present.
func (t *LocalFileTransport) ensureGitignore() {
	// Try to find .gitignore relative to baseDir's parent
	gitignorePath := filepath.Join(filepath.Dir(t.baseDir), ".gitignore")

	content, err := os.ReadFile(gitignorePath)
	if err != nil {
		content = []byte{}
	}

	if !strings.Contains(string(content), ".uncaught") {
		line := "\n# Uncaught local error store\n.uncaught/\n"
		f, err := os.OpenFile(gitignorePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
		if err != nil {
			return // Non-critical
		}
		defer f.Close()
		f.WriteString(line)
	}
}

// updateIssuesIndex reads, updates, and atomically writes the issues.json index.
func (t *LocalFileTransport) updateIssuesIndex(event *UncaughtEvent, eventFile, promptFile string) error {
	indexPath := filepath.Join(t.baseDir, "issues.json")

	// Read existing issues
	var issues []IssueEntry
	data, err := os.ReadFile(indexPath)
	if err == nil {
		json.Unmarshal(data, &issues) // Ignore errors — start fresh if malformed
	}

	// Determine user ID
	userID := "anonymous"
	if event.User != nil {
		if event.User.ID != "" {
			userID = event.User.ID
		} else if event.User.Email != "" {
			userID = event.User.Email
		}
	}

	// Find existing issue for this fingerprint
	existingIdx := -1
	for i, issue := range issues {
		if issue.Fingerprint == event.Fingerprint {
			existingIdx = i
			break
		}
	}

	if existingIdx >= 0 {
		// Update existing issue
		existing := &issues[existingIdx]
		existing.Count++
		existing.LastSeen = event.Timestamp
		existing.LatestEventFile = eventFile
		existing.FixPromptFile = promptFile

		// Add user if not already present
		found := false
		for _, u := range existing.AffectedUsers {
			if u == userID {
				found = true
				break
			}
		}
		if !found {
			existing.AffectedUsers = append(existing.AffectedUsers, userID)
		}

		// Re-open if previously resolved
		if existing.Status == Resolved {
			existing.Status = Open
		}
	} else {
		// Create new issue entry
		env := ""
		if event.Environment != nil {
			env = event.Environment.Deploy
		}
		issues = append(issues, IssueEntry{
			Fingerprint:     event.Fingerprint,
			Title:           event.Error.Message,
			ErrorType:       event.Error.Type,
			Count:           1,
			AffectedUsers:   []string{userID},
			FirstSeen:       event.Timestamp,
			LastSeen:        event.Timestamp,
			Status:          Open,
			FixPromptFile:   promptFile,
			LatestEventFile: eventFile,
			Release:         event.Release,
			Environment:     env,
		})
	}

	// Atomic write
	issuesJSON, err := json.MarshalIndent(issues, "", "  ")
	if err != nil {
		return err
	}
	return atomicWrite(indexPath, issuesJSON)
}

// atomicWrite writes data to a file atomically using tmp + rename.
func atomicWrite(path string, data []byte) error {
	tmpPath := path + ".tmp"
	if err := os.WriteFile(tmpPath, data, 0644); err != nil {
		return err
	}
	return os.Rename(tmpPath, path)
}

// ConsoleTransport logs events to stderr (useful for debugging).
type ConsoleTransport struct{}

// NewConsoleTransport creates a new console transport.
func NewConsoleTransport() *ConsoleTransport {
	return &ConsoleTransport{}
}

// Send logs the event to stderr.
func (t *ConsoleTransport) Send(event *UncaughtEvent) error {
	fmt.Fprintf(os.Stderr, "[uncaught] %s: %s\n", event.Error.Type, event.Error.Message)
	fmt.Fprintf(os.Stderr, "  Event ID: %s\n", event.EventID)
	fmt.Fprintf(os.Stderr, "  Fingerprint: %s\n", event.Fingerprint)
	if event.Error.Stack != "" {
		fmt.Fprintf(os.Stderr, "  Stack:\n%s\n", event.Error.Stack)
	}
	if event.FixPrompt != "" {
		fmt.Fprintf(os.Stderr, "  Fix Prompt:\n%s\n", event.FixPrompt)
	}
	return nil
}

// Flush is a no-op for the console transport.
func (t *ConsoleTransport) Flush() error {
	return nil
}

// createTransport creates the appropriate transport based on config.
func createTransport(config *Config) Transport {
	mode := config.Transport
	if mode == "" {
		mode = TransportLocal
	}

	switch mode {
	case TransportConsole:
		return NewConsoleTransport()
	case TransportLocal:
		return NewLocalFileTransport(config.LocalOutputDir)
	default:
		return NewLocalFileTransport(config.LocalOutputDir)
	}
}

File: packages/rust/src/transport.rs
// ---------------------------------------------------------------------------
// uncaught — transport layer (console / local-file)
// ---------------------------------------------------------------------------

use std::fs;
use std::path::{Path, PathBuf};

use crate::types::*;

/// A transport implementation capable of delivering events.
pub trait Transport {
    fn send(&self, event: &UncaughtEvent);
    fn flush(&self);
}

/// Create the appropriate transport strategy based on config.
pub fn create_transport(config: &UncaughtConfig) -> Box<dyn Transport + Send + Sync> {
    match config.transport {
        TransportMode::Console => Box::new(ConsoleTransport),
        TransportMode::Remote => Box::new(ConsoleTransport), // TODO: implement remote
        TransportMode::Local => Box::new(LocalFileTransport::new(config)),
    }
}

// ===================================================================
// Console Transport
// ===================================================================

struct ConsoleTransport;

impl Transport for ConsoleTransport {
    fn send(&self, event: &UncaughtEvent) {
        let title = format!(
            "[uncaught] {}: {}",
            event.error.error_type, event.error.message
        );
        eprintln!("--- {} ---", title);
        eprintln!("Event ID: {}", event.event_id);
        eprintln!("Fingerprint: {}", event.fingerprint);
        if let Some(ref stack) = event.error.stack {
            eprintln!("Stack: {}", stack);
        }
        if !event.fix_prompt.is_empty() {
            eprintln!("Fix Prompt:\n{}", event.fix_prompt);
        }
        eprintln!("---");
    }

    fn flush(&self) {
        // Nothing to flush for console transport.
    }
}

// ===================================================================
// Local File Transport
// ===================================================================

struct LocalFileTransport {
    base_dir: PathBuf,
}

impl LocalFileTransport {
    fn new(config: &UncaughtConfig) -> Self {
        let base_dir = match &config.local_output_dir {
            Some(dir) => PathBuf::from(dir),
            None => {
                let cwd = std::env::current_dir().unwrap_or_else(|_| PathBuf::from("."));
                cwd.join(".uncaught")
            }
        };

        // Ensure directory structure exists
        let _ = fs::create_dir_all(base_dir.join("events"));
        let _ = fs::create_dir_all(base_dir.join("fix-prompts"));

        // Auto-add .uncaught/ to .gitignore
        Self::ensure_gitignore(&base_dir);

        Self { base_dir }
    }

    fn ensure_gitignore(base_dir: &Path) {
        if let Some(parent) = base_dir.parent() {
            let gitignore_path = parent.join(".gitignore");
            let content = fs::read_to_string(&gitignore_path).unwrap_or_default();
            if !content.contains(".uncaught") {
                let line = "\n# Uncaught local error store\n.uncaught/\n";
                let _ = fs::write(
                    &gitignore_path,
                    format!("{}{}", content, line),
                );
            }
        }
    }
}

impl Transport for LocalFileTransport {
    fn send(&self, event: &UncaughtEvent) {
        let fp = &event.fingerprint;
        let event_dir = self.base_dir.join("events").join(fp);
        let _ = fs::create_dir_all(&event_dir);

        // Serialize event
        let json = match serde_json::to_string_pretty(event) {
            Ok(j) => j,
            Err(_) => return,
        };

        // Write timestamped event file (atomic: .tmp -> rename)
        let ts = event.timestamp.replace([':', '.'], "-");
        let event_file = format!("event-{}.json", ts);
        let event_path = event_dir.join(&event_file);
        let tmp_event_path = event_path.with_extension("json.tmp");
        if fs::write(&tmp_event_path, &json).is_ok() {
            let _ = fs::rename(&tmp_event_path, &event_path);
        }

        // Write / overwrite latest.json
        let latest_path = event_dir.join("latest.json");
        let tmp_latest_path = latest_path.with_extension("json.tmp");
        if fs::write(&tmp_latest_path, &json).is_ok() {
            let _ = fs::rename(&tmp_latest_path, &latest_path);
        }

        // Write fix-prompt Markdown file
        let prompt_file = format!("{}.md", fp);
        let prompt_path = self.base_dir.join("fix-prompts").join(&prompt_file);
        let tmp_prompt_path = prompt_path.with_extension("md.tmp");
        if fs::write(&tmp_prompt_path, &event.fix_prompt).is_ok() {
            let _ = fs::rename(&tmp_prompt_path, &prompt_path);
        }

        // Update issues.json index
        self.update_issues_index(event, &event_file, &prompt_file);
    }

    fn flush(&self) {
        // Local file transport writes synchronously per-event; nothing to flush.
    }
}

impl LocalFileTransport {
    fn update_issues_index(&self, event: &UncaughtEvent, event_file: &str, prompt_file: &str) {
        let index_path = self.base_dir.join("issues.json");

        let mut issues: Vec<IssueEntry> = fs::read_to_string(&index_path)
            .ok()
            .and_then(|raw| serde_json::from_str(&raw).ok())
            .unwrap_or_default();

        let user_id = event
            .user
            .as_ref()
            .and_then(|u| u.id.clone().or(u.email.clone()))
            .unwrap_or_else(|| "anonymous".to_string());

        if let Some(existing) = issues.iter_mut().find(|i| i.fingerprint == event.fingerprint) {
            existing.count += 1;
            existing.last_seen = event.timestamp.clone();
            existing.latest_event_file = event_file.to_string();
            existing.fix_prompt_file = prompt_file.to_string();
            if !existing.affected_users.contains(&user_id) {
                existing.affected_users.push(user_id);
            }
            if existing.status == IssueStatus::Resolved {
                existing.status = IssueStatus::Open;
            }
        } else {
            issues.push(IssueEntry {
                fingerprint: event.fingerprint.clone(),
                title: event.error.message.clone(),
                error_type: event.error.error_type.clone(),
                count: 1,
                affected_users: vec![user_id],
                first_seen: event.timestamp.clone(),
                last_seen: event.timestamp.clone(),
                status: IssueStatus::Open,
                fix_prompt_file: prompt_file.to_string(),
                latest_event_file: event_file.to_string(),
                release: event.release.clone(),
                environment: event.environment.as_ref().and_then(|e| e.deploy.clone()),
            });
        }

        // Atomic write
        let tmp_index_path = index_path.with_extension("json.tmp");
        if let Ok(json) = serde_json::to_string_pretty(&issues) {
            if fs::write(&tmp_index_path, &json).is_ok() {
                let _ = fs::rename(&tmp_index_path, &index_path);
            }
        }
    }
}
